### PR TITLE
[C#] Checkpoint and benchmark improvements + tests

### PR DESCRIPTION
**Pull request by badrishc**
* Update FASTER.benchmark with more options for frequent checkpoints
* Optimize checkpointing time by adding PreallocateLog option to LogSettings
* Improve sync/async code paths in checkpoint state machine
* Fixed bugs in checkpointing logic when fast-forwarding stale client sessions
* Make checkpointing state machine sync-only, corralling async loop to `CompleteCheckpointAsync`
* Added new state machine test suite to verify state machine behavior

Fix https://github.com/microsoft/FASTER/issues/295
Fix https://github.com/microsoft/FASTER/issues/296

### CODE REVIEW

File: cs/src/core/Index/FASTER/FASTER.cs [hunk 1]
```diff
@@ -345,9 +346,18 @@ public async ValueTask CompleteCheckpointAsync(CancellationToken token = default
                     systemState.phase == Phase.IN_PROGRESS_GROW)
                     return;
 
-                // TODO: Do we need access to functions here?
-                // If yes then move this to either faster legacy or client session.
-                await ThreadStateMachineStep<Empty, Empty, Empty, NullFasterSession>(null, NullFasterSession.Instance, true, token);
+                List<ValueTask> valueTasks = new List<ValueTask>();
+                
+                ThreadStateMachineStep<Empty, Empty, Empty, NullFasterSession>(null, NullFasterSession.Instance, ref valueTasks, token);
```
Comment (tli2) [inline cs/src/core/Index/FASTER/FASTER.cs:351]: This `ref` seems unnecessary

File: cs/src/core/Index/FASTER/FASTERThread.cs [hunk 2]
```diff
@@ -27,7 +27,7 @@ public partial class FasterKV<Key, Value> : FasterBase, IFasterKV<Key, Value>
                     var currentState = SystemState.Copy(ref systemState);
                     if (currentState.phase == Phase.REST)
                     {
-                        var intermediateState = SystemState.Make(Phase.INTERMEDIATE, currentState.version);
+                        var intermediateState = SystemState.Make(currentState.phase | Phase.INTERMEDIATE, currentState.version);
```
Comment (tli2) [inline cs/src/core/Index/FASTER/FASTERThread.cs:30]: Maybe refactor this into a method for readability?

Comment (badrishc) [reply cs/src/core/Index/FASTER/FASTERThread.cs]: Done

File: cs/src/core/Index/FASTER/FASTER.cs (hunk 1 above)
Comment (badrishc) [reply cs/src/core/Index/FASTER/FASTER.cs]: Removed the ref.

**Comment by tli2** [comment]
This only partially fixes #295 , right?

**Comment by badrishc** [comment]
There should no longer be any correctness issue. It's possible for a thread to wait on a future flush, but this PR ensures forward progress. A way to fix it better would be to re-check whether thread local state is in a cycle prior to the current global state, after reading the semaphore.

**Comment by badrishc** [comment]
Latest check-in fixes the latency issue by preventing thread waits on future checkpoint cycles.

**Pull request merged**